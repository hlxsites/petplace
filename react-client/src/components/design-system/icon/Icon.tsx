import { classNames } from "~/util/styleUtil";
import AddIcon from "./assets/add-icon.svg";
import AlertIcon from "./assets/alert-diamond-icon.svg";
import AppsIcon from "./assets/apps-icon.svg";
import BendArrowDownIcon from "./assets/bend-arrow-down.svg";
import CheckIcon from "./assets/check-icon.svg";
import ChevronLeftIcon from "./assets/chevron-left-icon.svg";
import CpuChipIcon from "./assets/cpu-chip-icon.svg";
import EllipseIcon from "./assets/ellipse-icon.svg";
import FileIcon from "./assets/file-icon.svg";
import MedicineIcon from "./assets/medicine-icon.svg";
import PawIcon from "./assets/paw-icon.svg";
import PippetIcon from "./assets/pippet-icon.svg";
import ShieldGoodIcon from "./assets/shield-good-icon.svg";
import ShieldOffIcon from "./assets/shield-off-icon.svg";
import SyringeIcon from "./assets/syringe-icon.svg";
import TransferIcon from "./assets/transfer-icon.svg";
import WarningTriangleIcon from "./assets/warning-triangle-icon.svg";

const IconMap = Object.freeze({
  add: <AddIcon />,
  alert: <AlertIcon />,
  apps: <AppsIcon />,
  bendArrowDown: <BendArrowDownIcon />,
  check: <CheckIcon />,
  chevronLeft: <ChevronLeftIcon />,
  cpuChip: <CpuChipIcon />,
  ellipse: <EllipseIcon />,
  file: <FileIcon />,
  medicine: <MedicineIcon />,
  paw: <PawIcon />,
  pippet: <PippetIcon />,
  shieldGood: <ShieldGoodIcon />,
  shieldOff: <ShieldOffIcon />,
  syringe: <SyringeIcon />,
  transfer: <TransferIcon />,
  warningTriangle: <WarningTriangleIcon />,
});

export type IconKeys = keyof typeof IconMap;

export type IconProps = {
  className?: string;
  display: IconKeys;
  size?: number;
};

export const Icon = ({ className, display, size, ...rest }: IconProps) => {
  return (
    <div
      className={classNames("icon w-base lg:w-large", className)}
      {...rest}
      style={{
        height: size,
        width: size,
      }}
    >
      {IconMap[display]}
    </div>
  );
};
