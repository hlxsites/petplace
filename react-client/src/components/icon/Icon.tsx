import CheckIcon from "./assets/check-icon.svg";

export const IconMap = Object.freeze({
  check: <CheckIcon />,
});

export type IconKeys = keyof typeof IconMap;

export type IconProps = {
  className?: string;
  color?: string;
  display: IconKeys;
  size?: number;
};

export const Icon = ({ color, display, size, ...rest }: IconProps) => {
  return (
    <div {...rest} style={{ color, width: size }}>
      {IconMap[display]}
    </div>
  );
};
