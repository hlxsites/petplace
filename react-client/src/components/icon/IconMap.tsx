import Ellipse from "./assets/ellipse-icon.svg";
import ShieldGood from "./assets/shield-good-icon.svg";
import ShieldOff from "./assets/shield-off-icon.svg";

const IconMap = {
  ellipse: <Ellipse />,
  shieldGood: <ShieldGood />,
  shieldOff: <ShieldOff />,
};

export default IconMap;
