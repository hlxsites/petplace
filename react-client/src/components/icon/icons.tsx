type IconProps = { size?: number; color?: string; fill: string };

export const IconCheck = ({ size = 20, color = "white", fill = "none" }: IconProps) => (
  <svg
    viewBox="0 0 20 20"
    fill={fill}
    xmlns="http://www.w3.org/2000/svg"
    width={size}
    height={size}
  >
    <g id="Checkbox">
      <rect id="Rectangle 2" width="20" height="20" rx="6" fill="#C74D2F" />
      <g id="Check">
        <path
          id="Vector"
          d="M14.6875 5.31213L8.39917 14.2951C8.3162 14.4143 8.20609 14.5121 8.07791 14.5804C7.94973 14.6487 7.80715 14.6856 7.66192 14.688C7.51669 14.6904 7.37297 14.6582 7.2426 14.5942C7.11223 14.5302 6.99895 14.436 6.91208 14.3196L5.3125 12.1871"
          stroke={color}
          strokeWidth="1.5"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </g>
    </g>
  </svg>
);

export const IconHeart = ({ size = 20, color = "white", fill = "none" }: IconProps) => (
  <svg
    viewBox="0 0 24 24"
    fill={fill}
    xmlns="http://www.w3.org/2000/svg"
    width={size}
    height={size}
  >
    <g id="Favorite">
      <path
        id="Vector"
        d="M12 21.844L2.41202 11.844C1.56935 11.002 1.01318 9.91606 0.82238 8.7402C0.631579 7.56434 0.815836 6.35825 1.34902 5.293C1.75108 4.48912 2.33839 3.79227 3.06257 3.25988C3.78674 2.72749 4.62706 2.37479 5.51427 2.23085C6.40149 2.0869 7.31021 2.15583 8.16557 2.43194C9.02092 2.70806 9.79843 3.18347 10.434 3.819L12 5.384L13.566 3.819C14.2016 3.18347 14.9791 2.70806 15.8345 2.43194C16.6898 2.15583 17.5985 2.0869 18.4858 2.23085C19.373 2.37479 20.2133 2.72749 20.9375 3.25988C21.6617 3.79227 22.249 4.48912 22.651 5.293C23.1835 6.35784 23.3674 7.56321 23.1768 8.7384C22.9862 9.91359 22.4307 10.999 21.589 11.841L12 21.844Z"
        stroke={color}
        strokeWidth="1.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </g>
  </svg>
);
