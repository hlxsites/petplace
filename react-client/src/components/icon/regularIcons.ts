import {
  faBell,
  faBuilding,
  faCalendar,
  faCalendarAlt,
  faCaretSquareDown,
  faCaretSquareUp,
  faCheckCircle,
  faCircle,
  faClipboard,
  faCopy,
  faDotCircle,
  faEnvelope,
  faEye,
  faEyeSlash,
  faFileAlt,
  faHeart,
  faImage,
  faLifeRing,
  faMinusSquare,
  faSave,
  faStar,
  faStickyNote,
  faTimesCircle,
  faTrashAlt,
  faUser,
  faUserCircle,
} from "@fortawesome/free-regular-svg-icons";

/**
 * naming convention:
 * (1) remove the "fa" prefix from the imported name
 * (2) add a "R" char at the end, to make it clear that is a regular-type icon
 * FontAwesome can have the same icon with the same name in different types, so make sure to name properly
 */
export const regularIcons = {
  bellR: faBell,
  buildingR: faBuilding,
  calendarAltR: faCalendarAlt,
  calendarR: faCalendar,
  caretSquareDownR: faCaretSquareDown,
  caretSquareUpR: faCaretSquareUp,
  checkCircleR: faCheckCircle,
  circleR: faCircle,
  clipboardR: faClipboard,
  copyR: faCopy,
  dotCircleR: faDotCircle,
  envelopeR: faEnvelope,
  eyeR: faEye,
  eyeSlashR: faEyeSlash,
  fileAltR: faFileAlt,
  heartR: faHeart,
  imageR: faImage,
  lifeRingR: faLifeRing,
  minusSquareR: faMinusSquare,
  saveR: faSave,
  starR: faStar,
  stickNoteR: faStickyNote,
  timesCircleR: faTimesCircle,
  trashAltR: faTrashAlt,
  userCircleR: faUserCircle,
  userR: faUser,
};
