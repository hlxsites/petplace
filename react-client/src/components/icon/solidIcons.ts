import {
  faAlignCenter,
  faAlignLeft,
  faAlignRight,
  faArrowUp,
  faArrowUpRightFromSquare,
  faBars,
  faBackward,
  faBicycle,
  faBold,
  faBolt,
  faBook,
  faBurger,
  faCamera,
  faCaretDown,
  faCartShopping,
  faCaretLeft,
  faCaretRight,
  faCarSide,
  faCheck,
  faCheckDouble,
  faCheckSquare,
  faChevronDown,
  faChevronLeft,
  faChevronRight,
  faChevronUp,
  faCog,
  faCross,
  faEllipsisH,
  faEllipsisV,
  faExclamation,
  faExclamationCircle,
  faFileInvoice,
  faFire,
  faFont,
  faForward,
  faGift,
  faHandHoldingHeart,
  faHeadset,
  faHighlighter,
  faHospital,
  faHouse,
  faInfoCircle,
  faItalic,
  faLink,
  faListOl,
  faListUl,
  faMinus,
  faMinusCircle,
  faMotorcycle,
  faPiggyBank,
  faPen,
  faPhone,
  faPlus,
  faPlane,
  faRedo,
  faShirt,
  faShield,
  faSlidersH,
  faSort,
  faSortDown,
  faSortUp,
  faStar,
  faTable,
  faTablet,
  faTimes,
  faTrash,
  faTriangleExclamation,
  faUnderline,
  faUndo,
  faUnlink,
  faUpload,
} from "@fortawesome/free-solid-svg-icons";

/**
 * naming convention:
 * (1) remove the "fa" prefix from the imported name
 * (2) add a "S" char at the end, to make it clear that is a solid-type icon
 * FontAwesome can have the same icon with the same name in different types, so make sure to name properly
 */
export const solidIcons = {
  alignCenterS: faAlignCenter,
  alignLeftS: faAlignLeft,
  alignRightS: faAlignRight,
  arrowUpRightFromSquareS: faArrowUpRightFromSquare,
  arrowUpS: faArrowUp,
  backwardS: faBackward,
  barsS: faBars,
  bicycleS: faBicycle,
  boldS: faBold,
  boltS: faBolt,
  bookS: faBook,
  burgerS: faBurger,
  caretDownS: faCaretDown,
  cameraS: faCamera,
  cartShoppingS: faCartShopping,
  caretLeftS: faCaretLeft,
  caretRightS: faCaretRight,
  carSideS: faCarSide,
  checkS: faCheck,
  checkDoubleS: faCheckDouble,
  checkSquareS: faCheckSquare,
  chevronDownS: faChevronDown,
  chevronLeftS: faChevronLeft,
  chevronRightS: faChevronRight,
  chevronUpS: faChevronUp,
  cogS: faCog,
  crossS: faCross,
  ellipsisHS: faEllipsisH,
  ellipsisVS: faEllipsisV,
  exclamationCircleS: faExclamationCircle,
  exclamationS: faExclamation,
  giftS: faGift,
  fireS: faFire,
  fileInvoiceS: faFileInvoice,
  fontS: faFont,
  forwardS: faForward,
  headsetS: faHeadset,
  handHoldingHeartS: faHandHoldingHeart,
  highlighterS: faHighlighter,
  hospitalS: faHospital,
  houseS: faHouse,
  infoCircleS: faInfoCircle,
  italicS: faItalic,
  linkS: faLink,
  listOlS: faListOl,
  listUlS: faListUl,
  minusCircleS: faMinusCircle,
  minusS: faMinus,
  motorcycleS: faMotorcycle,
  penS: faPen,
  phoneS: faPhone,
  piggyBankS: faPiggyBank,
  plusS: faPlus,
  planeS: faPlane,
  redoS: faRedo,
  slidersHS: faSlidersH,
  shirtS: faShirt,
  shieldS: faShield,
  sortDownS: faSortDown,
  sortS: faSort,
  sortUpS: faSortUp,
  starS: faStar,
  tableS: faTable,
  tabletS: faTablet,
  timesS: faTimes,
  trashS: faTrash,
  triangleExclamationS: faTriangleExclamation,
  underlineS: faUnderline,
  undoS: faUndo,
  unlinkS: faUnlink,
  uploadS: faUpload,
};
