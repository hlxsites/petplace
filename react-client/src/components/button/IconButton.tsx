import { forwardRef } from "react";
import { classNames } from "../../util/styleUtil";
import { Icon, type IconKeys, type IconProps } from "../icon/Icon";
import useButtonBase, { type IUseButtonBase } from "./useButtonBase";

export type IconButtonProps = IUseButtonBase &
  Omit<JSX.IntrinsicElements["button"], "children"> & {
    label: string;
    icon: IconKeys;
    iconProps?: Omit<IconProps, "display">;
    type?: "button" | "submit" | "reset";
    onClick?: () => void;
  };

const IconButton = forwardRef<HTMLButtonElement, IconButtonProps>(
  (
    {
      className,
      icon,
      iconProps,
      isFullWidth,
      isLoading,
      label,
      type = "button",
      variant,
      ...rest
    }: IconButtonProps,
    ref
  ) => {
    const { className: baseClassName } = useButtonBase({
      isFullWidth,
      isLoading,
      variant,
    });

    return (
      <button
        aria-label={label}
        className={classNames(
          baseClassName,
          "btn-square py-0 mx-1 min-w-0 px-[12px]",
          className
        )}
        type={type}
        ref={ref}
        {...rest}
      >
        <Icon
          display={icon}
          {...iconProps}
          style={{ maxHeight: "100%", maxWidth: "100%" }}
        />
      </button>
    );
  }
);

IconButton.displayName = "IconButton";

export default IconButton;
