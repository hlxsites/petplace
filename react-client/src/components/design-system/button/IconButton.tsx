import { forwardRef } from "react";
import { classNames } from "~/util/styleUtil";
import { Icon, IconProps } from "../icon/Icon";
import useButtonBase, { type UseButtonBase } from "./useButtonBase";

export type IconButtonProps = UseButtonBase &
  Omit<JSX.IntrinsicElements["button"], "children" | "fullWidth"> & {
    label: string;
    icon: IconProps["display"];
    iconProps?: Omit<IconProps, "display">;
    type?: "button" | "submit" | "reset";
    onClick?: () => void;
  };

export const IconButton = forwardRef<HTMLButtonElement, IconButtonProps>(
  (
    {
      className,
      icon,
      iconProps,
      isLoading,
      label,
      type = "button",
      variant,
      ...rest
    }: IconButtonProps,
    ref
  ) => {
    const { className: baseClassName } = useButtonBase({
      isLoading,
      variant,
    });

    return (
      <button
        aria-label={label}
        className={classNames(
          baseClassName,
          "btn-square mx-1 min-w-0 px-medium py-0",
          className
        )}
        type={type}
        ref={ref}
        {...rest}
      >
        <Icon className="max-h-full max-w-full" display={icon} {...iconProps} />
      </button>
    );
  }
);
