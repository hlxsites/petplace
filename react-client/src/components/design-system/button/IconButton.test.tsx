import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ComponentProps } from "react";

import { IconKeys } from "../icon/Icon";
import { IconButton } from "./IconButton";

const { getByRole, getByLabelText } = screen;

describe("<IconButton />", () => {
  it.each(["a-class", "another-class"])(
    `should render button with custom className '%s'`,
    (expected) => {
      getRenderer({ className: expected });
      expect(getByRole("button")).toHaveClass(expected);
    }
  );

  it.each(["Label", "Another label"])(
    "should render button with label '%s'",
    (expected) => {
      getRenderer({ label: expected });
      expect(getByLabelText(expected).tagName).toBe("BUTTON");
    }
  );

  it.each(["Add", "Check"])(
    `should render an button with '%s' icon`,
    (expected) => {
      const { container } = getRenderer({
        icon: expected.toLowerCase() as IconKeys,
      });
      expect(container.querySelector("svg")).toHaveAttribute(
        "data-file-name",
        `Svg${expected}Icon`
      );
    }
  );

  it("should call onClick callback", async () => {
    const onClick = jest.fn();

    getRenderer({ onClick });
    expect(onClick).not.toHaveBeenCalled();
    const button = getByRole("button");

    await userEvent.click(button);
    await userEvent.click(button);
    expect(onClick).toHaveBeenCalledTimes(2);
  });

  it("should have type='button' by default", () => {
    getRenderer();
    expect(getByRole("button")).toHaveAttribute("type", "button");
  });

  it("should have type='submit", () => {
    getRenderer({ type: "submit" });
    expect(getByRole("button")).toHaveAttribute("type", "submit");
  });

  it("should be enabled by default", () => {
    getRenderer();
    expect(getByRole("button")).toBeEnabled();
  });

  it("should be disabled", () => {
    getRenderer({ disabled: true });
    expect(getByRole("button")).toBeDisabled();
  });

  it.each(["sample-id", "another-id"])(
    "button should have id '%s'",
    (expected) => {
      getRenderer({ id: expected });
      expect(getByRole("button")).toHaveAttribute("id", expected);
    }
  );

  it.each(["Custom button", "Another button"])(
    "should have aria-label '%s'",
    (expected) => {
      getRenderer({ "aria-label": expected });
      expect(getByRole("button")).toHaveAttribute("aria-label", expected);
    }
  );

  it('should render button with "primary" variant by default', () => {
    getRenderer();
    expect(getByRole("button")).toHaveClass(
      "bg-orange-300-contrast text-white"
    );
  });

  it('should render button with "secondary" variant', () => {
    getRenderer({ variant: "secondary" });
    expect(getByRole("button")).toHaveClass("text-neutral-700");
  });

  it('should render button with "link" variant', () => {
    getRenderer({ variant: "link" });
    expect(getByRole("button")).toHaveClass(
      "bg-transparent border-0 text-neutral-700"
    );
  });

  it('should render button with "error" variant', () => {
    getRenderer({ variant: "error" });
    expect(getByRole("button")).toHaveClass("bg-red-300 text-white");
  });

  it("should not be loading by default", () => {
    getRenderer();
    expect(getByRole("button")).not.toHaveClass("loading");
  });

  it("should render button with loading state", () => {
    getRenderer({ isLoading: true });
    expect(getByRole("button")).toHaveClass("loading");
  });
});

// Helpers
type Props = ComponentProps<typeof IconButton>;
function getRenderer({
  icon = "add",
  label = "Test label",
  ...rest
}: Partial<Props> = {}) {
  return render(<IconButton icon={icon} label={label} {...rest} />);
}
