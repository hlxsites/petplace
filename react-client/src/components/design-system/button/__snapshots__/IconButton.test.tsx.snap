// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> should match snapshot to assure that the component is being rendered correctly 1`] = `
<div>
  <button
    aria-label="Test label"
    class="rounded-full border-1 font-semibold outline outline-transparent outline-2 outline-offset-2 items-center flex justify-center px-base py-small lg:px-large text-base lg:py-[12px] bg-orange-300-contrast text-white hover:bg-orange-500 focus:bg-orange-300-contrast focus:outline-orange-100 btn-square py-0 mx-1 min-w-0 px-[12px]"
    type="button"
  >
    <div
      class="icon max-h-full max-w-full"
    >
      <svg
        data-file-name="SvgAddIcon"
      />
    </div>
  </button>
</div>
`;
